### DISCUSSION TITLE
STM32H7 CDC buffer size increase

### DISCUSSION THREAD

**Discussion by BennyEvans** [hathach/tinyusb#979] [Q&A]
Hi,

I'm still playing around with HS CDC serial on a waveshare STM32H7. I'm trying to increase throughput and have made the following modifications to the CDC buffers:

```
// CDC FIFO size of TX and RX
#define CFG_TUD_CDC_RX_BUFSIZE   (TUD_OPT_HIGH_SPEED ? 2048 : 64)
#define CFG_TUD_CDC_TX_BUFSIZE   (TUD_OPT_HIGH_SPEED ? 2048 : 64)

// CDC Endpoint transfer buffer size, more is faster
#define CFG_TUD_CDC_EP_BUFSIZE   (TUD_OPT_HIGH_SPEED ? 2048 : 64)
```

This seems to work and in a throughput test where the embedded device is sending data as fast as it can, it does seem to increase throughput. However, there seems to be an issue with these settings when receiving data. When sending data from a PC of exactly 512, 1024 or 1536 bytes in length, I don't get a response nor do I see the data on the chip with a call to `tud_cdc_available()`. The HS interrupt does fire and the RxFIFO nonempty bit is set. The difference that I see between a transfer of one of the problem lengths and any other length is that in any other length, the loop in dcd_synopsys.c calling `handle_rxflvl_ints ` (link below this paragraph) is entered a second time with the second packet status being 'Out packet done'. The load of `usb_otg->GINTSTS` then has the OUT endpoint interrupt bit set and the `handle_epout_ints` function is entered. On a problem length transfer, `handle_rxflvl_ints` is only entered once and not a second time with packet status being 'Out packet done'. Again, I'm still not up to speed with the internals of USB so what it's doing may be correct, but I'm not entirely sure what the issue is.

https://github.com/hathach/tinyusb/blob/b6b5a33806527ed06b57930067eb4a386839d274/src/portable/st/synopsys/dcd_synopsys.c#L1116

Unfortunately there is no log messages to assist as nothing is logged when sending data of one of these lengths. If someone is able to confirm this issue or have idea of what it could be, it'd be a great help.

Thanks,
Ben

**Comment by hathach** [comment]
@BennyEvans how do you send exactly 512 bytes ?  Normally if host driver send exactly a packet size, it will also send an zero length packet (ZLP) if there is no more data. Which is not your test case here, therefore tinyusb cdc driver continue to wait for next packet and does not attempt to notify application yet.

**Reply by BennyEvans** [reply]
@hathach I'm using this python script to test this issue:
```
import serial
import array

PORT = 'COM10'

def compare_array(orig, new, bytes):
    for i in range(0, bytes):
        if (orig[i] != new[i]):
            return False
    return True

def rx_and_compare(ser, numBytes):
    data = b""
    while len(data) < numBytes:
        if ser.inWaiting() > 0:
            readdata = ser.read(ser.inWaiting())
            data = data + readdata
    if compare_array(dataArray, data, numBytes):
        print(str(numBytes) + " match!")
    else:
        print(str(numBytes) + " no match!")
    return

with serial.Serial(PORT, writeTimeout = 0) as ser:
    try:
        if ser.is_open:
            dataArray = []
            #create data array
            for i in range(0, 2048):
                num = i % 256
                if num == 0:
                    num = 1
                dataArray.append(num)

            #Transfer 511 bytes - OK
            numBytes = 511
            toSend = array.array('B', dataArray[:numBytes]).tostring()
            ser.write(toSend)
            rx_and_compare(ser, numBytes)

            #Transfer 512 bytes - NOT OK
            numBytes = 512
            toSend = array.array('B', dataArray[:numBytes]).tostring()
            ser.write(toSend)
            rx_and_compare(ser, numBytes)

            print("Test complete!")
        else:
            print("Could not open port!")
    except:
        print("Error!")
    print("Closing port!")
    ser.close()
```
It sends 511 bytes of data and then 512 bytes and compares the result each time. As disucssed in the OP, when using a buffer size of 2048, I do not get a response from the 512 byte transfer. When using the default buffer size of 512 bytes, I do get a response.

I've captured the exchange using Wireshark detailed below:

**With 2048 byte buffers**

```
// CDC FIFO size of TX and RX
#define CFG_TUD_CDC_RX_BUFSIZE   (TUD_OPT_HIGH_SPEED ? 2048 : 64)
#define CFG_TUD_CDC_TX_BUFSIZE   (TUD_OPT_HIGH_SPEED ? 2048 : 64)

// CDC Endpoint transfer buffer size, more is faster
#define CFG_TUD_CDC_EP_BUFSIZE   (TUD_OPT_HIGH_SPEED ? 2048 : 64)
```

Pcap file: [511_512_capture_2048B_buffer.zip](https://github.com/hathach/tinyusb/files/6883451/511_512_capture_2048B_buffer.zip)

It looks like a zero byte packet is sent but it not processed by the device? I'm not sure.

**With 512 byte buffers**

```
// CDC FIFO size of TX and RX
#define CFG_TUD_CDC_RX_BUFSIZE   (TUD_OPT_HIGH_SPEED ? 512 : 64)
#define CFG_TUD_CDC_TX_BUFSIZE   (TUD_OPT_HIGH_SPEED ? 512: 64)

// CDC Endpoint transfer buffer size, more is faster
#define CFG_TUD_CDC_EP_BUFSIZE   (TUD_OPT_HIGH_SPEED ? 512: 64)
```

Pcap file: [511_512_capture_512B_buffer.zip](https://github.com/hathach/tinyusb/files/6883460/511_512_capture_512B_buffer.zip)

If there is a better format to upload the pcap files, let me know. It wouldn't let me upload them directly, only within a zip file.

Thanks,
Ben

**Reply by hathach** [reply]
@BennyEvans as mentioned above, when sending exactly the endpoint size, and there is no more data, host should send an extra ZLP to let device know the transfer is complete. In above case, you need to
- send 512 byte
- send 0 byte

Otherwise there is no way device could know if there is upcoming data e.g like transferring bulk of 4096 bytes. Which is the whole point to increased the EP_BUFSZIE to 2048 so that it could hold more than 1 packet per transfer and save time for stack to processing the data.

**Reply by HiFiPhile** [reply]
@hathach 
It's good to know ZLP need to be handled by application. I thought it has been handled by the driver 🤣

**Reply by hathach** [reply]
@HiFiPhile it totally depends on the host platform and the driver as well. We should probably separate the TX & RX BUFSIZE. driver such as MSC does not need ZLP since both host & device agree on how many they would transfer beforehand (via command packet).

**Reply by BennyEvans** [reply]
@hathach Thanks for the response. Have you had a look at the pcap files I linked in the post? I believe a ZLP is sent after the 512byte transfer - potentially handled by the python lib. I'm not at the computer at the moment so can't investigate further at the moment though.

I'm still a bit confused by it. Really appreciate your help though.

**Reply by hathach** [reply]
@BennyEvans oh yeah indeed, zlp is sent according to pcap file. It probably a bug with cdc or dcd driver. Sorry, I haven't read your response carefully enough. I will try to reproduce it with your python script and verify and fix it if needed.

**Reply by BennyEvans** [reply]
@hathach great, thanks. If you can verify its happening for you too, that'd be great. I'll keep looking into the issue but still wrapping my head around it all.

**Reply by BennyEvans** [reply]
I've been trying to debug this but haven't found anything obvious yet. It's worth noting though that the issue seems dependent on `CFG_TUD_CDC_EP_BUFSIZE`. For instance, the configuration below with `CFG_TUD_CDC_RX_BUFSIZE` and `CFG_TUD_CDC_TX_BUFSIZE` set to 2048 works fine for the 512 byte transfer but as soon as `CFG_TUD_CDC_EP_BUFSIZE` is increased, the error becomes apparent.

```
// ** Working configuration **

// CDC FIFO size of TX and RX
#define CFG_TUD_CDC_RX_BUFSIZE   (TUD_OPT_HIGH_SPEED ? 2048 : 64)
#define CFG_TUD_CDC_TX_BUFSIZE   (TUD_OPT_HIGH_SPEED ? 2048 : 64)

// Increasing this introduces the error
// CDC Endpoint transfer buffer size, more is faster
#define CFG_TUD_CDC_EP_BUFSIZE   (TUD_OPT_HIGH_SPEED ? 512 : 64)
```

Should I change this post to an issue to keep it logged there?

Cheers,
Ben

**Reply by HiFiPhile** [reply]
I observed the same effect with LPC4337 and STM32L072.

> I believe a ZLP is sent after the 512byte transfer 

If you look at the direction of ZLP packet it's device->host, so in fact host didn't sent the ZLP.

That's explains:
https://social.msdn.microsoft.com/Forums/en-US/920968a9-c52c-4ece-9c09-f368fbe46fe9/issue-with-usb-drivers-usbser-usbccid-regarding-zlp?forum=wdk

But I can't find how to send ZLP on Windows...

**Reply by BennyEvans** [reply]
@HiFiPhile ahhhh, I didn't notice that, thanks for pointing it out. Yeah, it's a little frustrating and doesn't seem to be documented well in many of the serial libraries, but I guess it does seem like its up to the app developer to send the ZLP. It makes things a little more interesting to implement.

I'm currently experimenting with bulk endpoints with a vendor class device so I assume I'm going to have the same issue there. It would be nice to have a flag in tinyUSB so that it always notifies the application as soon as data is available, though this may defeat the point.

I'll do a bit more digging into it tomorrow.
